new functions oneOf : List (Maybe a) -> a and oneOfLazy

skyqrose opened the pull request:
Alternative to #47

TSFoster commented:
Regardless of what happens to #47, if this version is decided to be a good addition, I wonder if there is a better name for it? `firstJust` or something? Because `oneOf` seems too tied to decoder/parser-like things.

These are (I think) all the `oneOf`s published:

<img width="225" alt="image" src="https://user-images.githubusercontent.com/510520/69067780-99dcca80-0a1b-11ea-9713-4e1d998fb58b.png">

If you squint, all but two (`Exts.Maybe.oneOf` and `Maybe017.oneOf` — both the same as this) are to do with building some sort of function to decode in some way (`Fuzz.oneOf` "decodes" a value from a `Seed`, right!?).

skyqrose commented:
Is there a common name for this kind of `List.foldl or` function, for other modules where `or` makes sense?

`Result.Extra` has `or`, but no `List (Result e a) -> Result e a`

Here's an interesting [package](https://package.elm-lang.org/packages/Janiczek/transform/1.1.0/Transform#orList) that works with decoder-like things and has `or : (a -> Maybe a) -> (a -> Maybe a) -> a -> Maybe a` and `orList : List (a -> Maybe a) -> a -> Maybe a` 

This [parsing package](https://package.elm-lang.org/packages/andre-dietrich/parser-combinators/3.2.0/Combine#choice) uses `choice : List (Parser s a) -> Parser s a` instead of `oneOf`.

From those examples, I'd lean towards `orList` and `orListLazy` if `oneOf` is used for the `List (a -> Maybe b)` version.

skyqrose commented:
After more discussion in #47, I've changed the implementation from `foldl` to recursion so it can short circuit and not need to scan the whole list, and I've renamed it to `orList` so that #47 can keep the name `oneOf`.
